Comment Detail: Content cell initial view

dvdchr opened the pull request:
Refs #17108, #17087
Depends on #17106

This is the first part of #17108, which adds the initial table view cell for comment content. Note that the cell is not yet populated by a Comment object, and therefore looks empty.

Some preview:

![Simulator Screen Shot - iPhone 8 - 2021-09-02 at 19 38 05](https://user-images.githubusercontent.com/1299411/131844979-3bb9547f-3f58-47f1-89b1-dfbca4dd4313.png)

Notes:
- I'm using SF Symbols for the button icons, since they are available to use on iOS 13.0.
- I noticed that the buttons (accessory, reply, like) have no highlighted state. I'll add them in a separate PR.
- In the next iteration, I'll populate the cell with a `Comment` object.

## To test

- Ensure the `New Comment Detail` feature flag is enabled.
- Go to My Site > Comments, and tap any comment.
- Verify that the comment content cell is displayed correctly:
  - The author's avatar image is displayed correctly.
  - The author's name is displayed correctly.
  - The date of the comment is displayed correctly.
  - The number of likes is displayed correctly.
  - The state of the liked button is correctly set (blue if liked, hollow with `.textSubtle` if unliked). Note that the like/unlike toggle is not implemented yet, so this is expected.

## Regression Notes
1. Potential unintended areas of impact
n/a. Feature is hidden behind a feature flag.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a. Feature is hidden behind a feature flag.

3. What automated tests I added (or what prevented me from doing so)
n/a. Feature is hidden behind a feature flag.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dvdchr committed:
Create the initial view for content cell

dvdchr committed:
Link the IBOutlets to the Swift class

dvdchr committed:
Show comment content cell in comment detail

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -0,0 +1,221 @@
+    // all the available images to display for the accessory button.
+    enum AccessoryButtonType {
+        case share
+        case threeDots
```
Can we call these `threeDots` names `ellipsis` instead?

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -0,0 +1,221 @@
+    // all the available images to display for the accessory button.
+    enum AccessoryButtonType {
+        case share
+        case threeDots
```
Can we change all the `threeDots` names to `ellipsis`?

dvdchr replied:
Sure. I defined the name before finding out that the name was actually `ellipsis` 😅 . This is addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17109/commits/8253b5d2977a170e21553f58b0123f90035dbbfd.

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -0,0 +1,221 @@
+
+    // MARK: - Public Properties
+
+    var displayName: String? = "" {
```
I'm assuming there will be a `configure` type method that will accept the `Comment`? That way all these `var`s can be set internally and thus can be private?

dvdchr replied:
Done in https://github.com/wordpress-mobile/WordPress-iOS/pull/17109/commits/05d5a31e08a13244e2a493f9cf2aea0b40f07f88.

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -0,0 +1,221 @@
+class CommentContentTableViewCell: UITableViewCell, NibReusable {
+
+    // determines the state of the like button.
+    enum LikeState {
```
Maybe I'm not seeing the big picture yet, but couldn't you just use `isLiked` from the `Comment`?

dvdchr replied:
Done in https://github.com/wordpress-mobile/WordPress-iOS/pull/17109/commits/05d5a31e08a13244e2a493f9cf2aea0b40f07f88.

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -0,0 +1,221 @@
+    @IBOutlet private weak var dateLabel: UILabel!
+    @IBOutlet private weak var accessoryButton: UIButton!
+
+    @IBOutlet weak var webView: WKWebView!
```
private?

dvdchr replied:
Done in https://github.com/wordpress-mobile/WordPress-iOS/pull/17109/commits/05d5a31e08a13244e2a493f9cf2aea0b40f07f88!

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -0,0 +1,221 @@
+        }
+    }
+
+    var displayDate: String? = "" {
```
It maybe be too early to nitpick these, but these empty strings confuse me. I'm not sure why they're defaulting to empty strings (instead of nil). But I also think once you populate them this will change anyway, so...

dvdchr replied:
Yeah, will update them later. The variable was initially `String`, but I figured later on it should just be an optional instead since `text` properties on the labels are optional strings too. ...But I forgot to update the default value for it. Thanks for noticing!

dvdchr replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17109/commits/05d5a31e08a13244e2a493f9cf2aea0b40f07f88. I removed the properties altogether since they're configured through a `Comment` object.

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -0,0 +1,221 @@
+
+    // MARK: Button Tap Handlers
+
+    var nameLabelTapAction: (() -> Void)? = nil
```
Maybe I missed it, but I don't see that tapping the user's name should do anything. Is this needed/what will this do?

dvdchr replied:
👋  ~~The current (old) comment detail shows the author's site in a modal window, so this is added for that.~~ Nevermind. There is the web address row for that. How did I not realize this. 😅   I'll remove this!

dvdchr replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17109/commits/d73aec748113eee66b945c205d05b8944e29b664!

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -0,0 +1,221 @@
+    // MARK: Component Visibility
+
+    // Hides the accessory button if the value is false.
+    var accessoryButtonEnabled: Bool = true {
```
All these button states can be private ya?

dvdchr replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17109/commits/05d5a31e08a13244e2a493f9cf2aea0b40f07f88!

ScoutHarris commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -0,0 +1,221 @@
+    /// Configures the avatar image view with the provided URL.
+    /// If the URL does not contain any image, the default placeholder image will be displayed.
+    /// - Parameter url: The URL containing the image.
+    func configureImage(with url: URL?) {
```
Again I'm assuming there will be a `configure` type method that will accept the `Comment`, which would call these image configure methods? Meaning they can be private?

dvdchr replied:
👍 . Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17109/commits/05d5a31e08a13244e2a493f9cf2aea0b40f07f88.

ScoutHarris commented:
It looks like the dark background on the webView shows sometimes:
- When a Comment is first selected. (iPad and iPhone)
- When the device is rotated. (seems like only iPad)


https://user-images.githubusercontent.com/1816888/131883741-d2613e73-b5e8-4ee1-b6b4-0d5efd317c57.mp4

dvdchr commented:
Hey @ScoutHarris 👋 . Thanks for the review!

> It looks like the dark background on the webView shows sometimes:

Thank you for the recordings! I intentionally left the default dark background to help with debugging the web content height, which will be addressed in the next PR. 😅 I'll remove it in this PR so it'll look more clean. 

 **Update**: This is addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17109/commits/e3d7bd9d5a13ed4e80bb474599837b033be544db ✅

> I'm assuming there will be a configure type method that will accept the Comment? That way all these vars can be set internally and thus can be private?
>
> Maybe I'm not seeing the big picture yet, but couldn't you just use isLiked from the Comment?
>
> All these button states can be private ya?
>
> Again I'm assuming there will be a configure type method that will accept the Comment, which would call these image configure methods? Meaning they can be private?

Previously, I considered whether to 1) hide all the view properties and just pass in a `Comment` object to configure everything privately, or 2) to expose the variables so it could be configured from the outside. Reason is, the cell may need to be configured with a `ReaderComment` in the future – this is referring to the comment threads down the road, since the content cells for the comment thread are likely to be reusing this too.

The implementation for this goes with (2), while you're _probably_ assuming (1) – hence why this raised some questions on your end. Nevertheless, I'll follow your suggestion to start simple with `configureWithComment` since the concerns that I've stated can be addressed later in the future. 🙂 Will update in a bit.

**Update**: 
- Most of the review comments above are addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17109/commits/05d5a31e08a13244e2a493f9cf2aea0b40f07f88 ✅. 
- Additionally, I've added `configure` method for the cell.
- I've also updated the testing instructions to verify which part of the cell should be displayed correctly.

dvdchr committed:
Update threeDots -> ellipsis

dvdchr committed:
Refactor comment content cell to be configured with Comment

dvdchr committed:
Set web view background color to transparent

dvdchr committed:
Improve docs

dvdchr committed:
Flip reply icon for RTL layout

dvdchr committed:
Remove tap gesture on name label

ScoutHarris commented:
> Reason is, the cell may need to be configured with a ReaderComment in the future – this is referring to the comment threads down the road, since the content cells for the comment thread are likely to be reusing this too.

Maybe I'm not seeing the problem, but I assumed a configure method would still suffice. In any case, thank you for clarifying. Clearly I assumed the opposite of your intent. And thanks for the changes!

ScoutHarris commented:
Unrelated to this PR (I think), but I noticed the changes from https://github.com/wordpress-mobile/WordPress-iOS/pull/17099/files#diff-fb74f65f40fd00682ceebb579df4be3d45796af24b8cf138ef0ba5915a44b8cb are not included here (update comment after editing). I just want to be sure it's not overwritten with your changes.

ScoutHarris reviewed (approved):
This is looking nice!

:shipit:

dvdchr commented:
Thanks @ScoutHarris! I'll merge the latest develop branch before merging this.

mokagio commented:
@dvdchr  I'm going to bump this to the next release because we'll be code freezing 18.2 today and this is a PR that is part of a WIP feature —we actually have a dedicated label for that "Part of WIP Feature".

The milestone is useful to help release train conductors [manage](https://mobile.blog/2021/03/12/two-simple-techniques-to-make-your-release-process-more-asynchronous-and-decentralized/) the distributed decision-making in the release process. When opening a PR that's part of a larger piece of work (which is a great idea, because small PRs are easy to review and iterate upon 😍😍😍) close to the end of a sprint, please use the milestone for the version after the upcoming one. Thanks! 🙇‍♂️

dvdchr commented:
> I'm going to bump this to the next release because we'll be code freezing 18.2 today

Noted, thank you for your help. 🙇 

> we actually have a dedicated label for that "Part of WIP Feature".

Oh, TIL! I'm definitely going to be using this label a lot. 🙂 

> When opening a PR that's part of a larger piece of work [...] close to the end of a sprint, please use the milestone for the version after the upcoming one. 

Noted, and will definitely keep in mind for future PRs. Thanks for reminding! 🙇

dvdchr committed:
Merge remote-tracking branch 'origin/develop' into issue/17087-content-cell
